### ISSUE TITLE
Feature Request: Compatibility with Ableton Live and CC MIDI Signals

### ISSUE THREAD

**Issue by someofthem98** [Logitech/logi_craft_sdk#29]
Dear Developers,
Could someone develop an extension for Ableton Live & a MIDI signal outpu for the crown of the Logitech Craft? It would be a killer feature for all the musicians & producers.
Many thanks in Advance

**Comment by SaveTheHuman5** [comment]
Take a look,
https://hi.computer/forum-community/talk-about-anything/logitech-craft-elephant/

**Comment by cyberluke** [comment]
> Take a look,
> https://hi.computer/forum-community/talk-about-anything/logitech-craft-elephant/

But this is for Surface Dial. I found your forum thread and in 2020 you wrote that Logitech Craft support will not be possible, right?

**Comment by marctdt** [comment]
> 
> 
> > Take a look,
> > https://hi.computer/forum-community/talk-about-anything/logitech-craft-elephant/
> 
> But this is for Surface Dial. I found your forum thread and in 2020 you wrote that Logitech Craft support will not be possible, right?

Hi @cyberluke 
I started to work on something for Ableton, so far I managed to make the crown scroll the browser. So I am curious to know why it is not possible? Is there some specific limitations?

**Comment by Flyweightt** [comment]
@cyberluke  I also own a Craft keyboard.
In the Logitec options software you can assign custom keys to the crown for scrolling..
But you need to enable developer mode.
Any change your software can support the crown form the craft keyboard?

**Comment by marctdt** [comment]
> @cyberluke I also own a Craft keyboard.
> In the Logitec options software you can assign custom keys to the crown for scrolling..
> But you need to enable developer mode.
> Any change your software can support the crown form the craft keyboard?

@Flyweightt  If you are interested I made some integration with Ableton. There 3 mode, Scroll browser, Mouse Wheel Mode simulation, and arrow button mode.
It was made for Ableton 11 Beta. I didn't update it yet for the suite version. If there some interest, I can maybe update it.
https://github.com/marctdt/LogicraftAbleton

**Comment by cyberluke** [comment]
@marctdt I'm on Github, so I'm developer. I google also a lot, I'm developer and googler. Already have developer mode, but only thing you can set is scrolling in 3rd party app, which I set to Arrow Up and Arrow Down. You cannot set any action for push and rotate. That is global action and you can map only like media keys or switch OS windows. The options are hardcoded.

I will check your extension. Can you please finish at least README.md? No problem with compiling stuff, but where to put it? Is it Ableton Python control surface script?

It would be great if you could navigate inside EQ plugin - navigate left and right for the knob and then change value. Because of this I think I will be buying Steinberg CC 121, which has awesome and complex Ableton support made by someone. This gives you channel strip functionality.

Right now I use Logicraft for changing volume or changing value inside a plugin. But the Ableton mixer is so poor. Reaper and Cubase are much better. It is difficult to read value on such tiny mixer. The best would be if you can put ballistic VU meter and also RMS read out values on something that is called Monitor channel in Reaper. That way you can put VST plugins that are processed automatically on the currently selected channel. Perhaps I can dig some Max4Live workaround script for that.

**Comment by marctdt** [comment]
> @marctdt I'm on Github, so I'm developer. I google also a lot, I'm developer and googler. Already have developer mode, but only thing you can set is scrolling in 3rd party app, which I set to Arrow Up and Arrow Down. You cannot set any action for push and rotate. That is global action and you can map only like media keys or switch OS windows. The options are hardcoded.
> 
> I will check your extension. Can you please finish at least README.md? No problem with compiling stuff, but where to put it? Is it Ableton Python control surface script?
> 
> It would be great if you could navigate inside EQ plugin - navigate left and right for the knob and then change value. Because of this I think I will be buying Steinberg CC 121, which has awesome and complex Ableton support made by someone. This gives you channel strip functionality.
> 
> Right now I use Logicraft for changing volume or changing value inside a plugin. But the Ableton mixer is so poor. Reaper and Cubase are much better. It is difficult to read value on such tiny mixer. The best would be if you can put ballistic VU meter and also RMS read out values on something that is called Monitor channel in Reaper. That way you can put VST plugins that are processed automatically on the currently selected channel. Perhaps I can dig some Max4Live workaround script for that.

That's correct without specific integration you cannot do much with the knob and the action of push and push+rotate is not accessible from the API, which is the shame. I thought of some workaround to use the push which could by sending some special character and reinterpret it on the agent. 

It's true that it is not really straight forward to install. I will try to give some instructions on the README.md . Basically it uses an agent made in .net to communicate with the craft restitute it to Ableton via some virtual midi port using the Python script. So you need to install a virtual midi port like loopMidi or BomeMidi Translator. Anyway the agent can also be used to simulate a mouse wheel to change parameters inside a plugin. I ve created some shortcut to switch from a mode to another.

If you need help I can try to guide you to install it. But notes that you will probably need to change some parameter as I made it with ableton 11 Beta. I will try to do that once I have freetime.

You can find the API for Ableton Python script here: https://structure-void.com/ableton-live-midi-remote-scripts/ .
I was looking for some ways to change mode by pointing or clicking some parameters in Ableton. But I am still not sure if it's possible

**Comment by cyberluke** [comment]
@marctdt Perfect, thank you! I have loopMidi. I think your solution is currently the best approach for Logitech Craft!

The information for install is enough for me, thank you. I may help when I'm less busy (next month). Perhaps we could use regular key stroke and translate it to midi or your app. So the crown would control midi value, but I could change midi cc number using some premapped keys. Lets say Q W E R T Y. Each key stroke would select a knob in EQ or channel strip and then you would use crown to change its value. The goal is to make it more like mixing console, have less screen time and use ears more.

**Comment by marctdt** [comment]
@cyberluke I will try to update that this week end.
Regarding sending midi CC, were you thinking of push+rotate? Unfortunately it doesn't seems to able to send Key stroke (see below)
![image](https://user-images.githubusercontent.com/38071131/110100014-3c03fb80-7da2-11eb-9d7e-38d14ad11df3.png)
Otherwise sending CC simple rotate should be possible.

Beside this, I think it may be possible to do what you suggested regarding using keyboard shortcut. The only issue is try to not make it conflict with other regular Ableton shortcut. We could also do double-tap or something else to change the eq parameter maybe.

**Comment by cyberluke** [comment]
@marctdt I was thinking regular key shortcut without Ableton conflict, yes. I don't know if Logitech Craft have some built-in support for this already. Bome Midi Translator can do such mapping. Problem here could be window focus, I think Logitech overlay should take care of this.

**Comment by marctdt** [comment]
@cyberluke I think we could send special character that are not really used and capture it in the agent. Then send it back to Ableton either via midi or directly capture the data in the Python script. 

I was going to give some instruction to install the agent. But I was wondering? Are you using windows or Mac? because I don't think it works on mac at this point.

**Comment by cyberluke** [comment]
I use Windows 10

**Comment by marctdt** [comment]
@cyberluke Ok, I will compile and package it to make it easier to install. I will also use the VirtualMidi SDK from loopMidi to automatically create a midi port.

**Comment by marctdt** [comment]
https://github.com/marctdt/LogicraftAbleton/releases/tag/v0.1
Here a first release.
I tried to give some instructions for the installation. Let me know if you have issues.

in terms of shortcut you can press:
alt + shift + ': to switch to browse mode
alt + shift + ,: to switch to mouse wheel mode
alt + shift + .: to switch to Arrow mode

The label and the icons of the overlay is from the sample code. I didn't change it :)

**Comment by cyberluke** [comment]
@marctdt Hi, this will not work. "Logitech Options v7.14.70" is impossible to use. Outdated and incompatible. I tried it, but it make major headaches with mouse Logitech, which use also this software layer. The zooming on Google Maps is completely crazy and there is a lot of other issues. You cannot make software for Ableton 11 and Windows 10 and then use like two years old Logitech runtime. That's a show stopper. With socket, you should communicate to Logitech about some exception for your profile or try to use something else. Perhaps forking the Python socket library, adding required options regarding socket or certificate and distribute with your app.

**Comment by marctdt** [comment]
@cyberluke Maybe I was not clear. Installing the version 7.14.70 is done as a workaround to install the profile in developper mode. Once this is done you can install the latest version of option. The communication with the socket library seems to remains the same in v7 or v8.
Please refer to some other ticket in this repo that refer to this issue. Unfortunately, it seems that Logitech is abandoning the Craft.

**Comment by cyberluke** [comment]
Ok, thank you. I can try it again. But the issues in your repo disappeared. I cannot see them even closed, it is empty here (even I change the filter settings): https://github.com/marctdt/LogicraftAbleton/issues ...I remember I was reading some info here. Even with 7.14.70, the directory structure is different. I tried to copy it in several locations and subdirectories, but it never appear as available profile to install. The testing profile from Craft SDK did appear and did install, though.

**Comment by marctdt** [comment]
> Ok, thank you. I can try it again. But the issues in your repo disappeared. I cannot see them even closed, it is empty here (even I change the filter settings): https://github.com/marctdt/LogicraftAbleton/issues ...I remember I was reading some info here. Even with 7.14.70, the directory structure is different. I tried to copy it in several locations and subdirectories, but it never appear as available profile to install. The testing profile from Craft SDK did appear and did install, though.

@cyberluke yes, i meant there some thread here that was saying that we cannot install custom profile from v8. I also tried many different location but i couldn’t make it work. So using the v7 is the only known workaround to install the profile at the moment. And I am afraid that moving forward they will stop support the craft.

**Comment by cyberluke** [comment]
Ok ok, cảm ơn nhiều :-) ...I could not even install custom profile in v7, it did not show up. I will try again.

**Comment by marctdt** [comment]
@cyberluke ahah, no worries. Make sure that it is in the right folder and that you are running ableton. Logitech options check the name of the process. What version of ableton do you use? 11 suite, standard, light ?

**Comment by cyberluke** [comment]
Ableton Live 11 Suite, but it shows in Logitech Options only as "Live". Through regedit I did check also GUID of Ableton app in Windows Programs, but then I realized your GUID is for your profile only, for the SDK, not for GUID of running app. So that should be okay, because they mismatch.

**Comment by marctdt** [comment]
@cyberluke yes the guid of the folder name is only related to SDK, not the guid of ableton. The options agent check the process name I think. The profile will appear in the list if the développer mode is activated and Ableton Live 11 Suite.exe process is running.

**Comment by marctdt** [comment]
If something is wrong with your setup, maybe check the manifest.json file. But I just looked at it and it should work if you installed properly the profile in the right folder, you activated the développer mode and run ableton live
